File: 01-Programming-Languages/Ruby/Algorithms/String/palindrome.rb
=begin
Problem: Valid Palindrome
Description: Check if a string reads the same forwards and backwards.

Approach:
- Use two pointers, one from start and one from end
- Compare characters while moving inward

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input:  "racecar"
Output: true

Input:  "hello"
Output: false
=end

def palindrome?(s)
  left = 0
  right = s.length - 1
  while left < right
    return false if s[left] != s[right]
    left += 1
    right -= 1
  end
  true
end

%w[racecar hello madam ruby].each do |s|
  puts "\"#{s}\" -> #{palindrome?(s)}"
end

File: 01-Programming-Languages/Dart/Algorithms/String/palindrome.dart
/*
Problem: Palindrome Check
Description: Check if a given string is a palindrome (reads the same
           forwards and backwards). Ignore non-alphanumeric characters
           and case.

Approach:
- Use two pointers (left, right) moving inward
- Skip non-alphanumeric characters
- Compare characters case-insensitively

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: "A man, a plan, a canal: Panama"
Output: true
*/

bool isPalindrome(String s) {
  String cleaned = s.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]'), '');
  int left = 0, right = cleaned.length - 1;
  while (left < right) {
    if (cleaned[left] != cleaned[right]) return false;
    left++;
    right--;
  }
  return true;
}

void main() {
  print("Palindrome 'A man, a plan, a canal: Panama': ${isPalindrome("A man, a plan, a canal: Panama")}");
  print("Palindrome 'race a car': ${isPalindrome("race a car")}");
}
